Tracks: add Woo identity

psealock opened the pull request:
### Changes proposed in this Pull Request:

Adds a Woo based identification for a user. Uses a Jetpack id if available, otherwise creates a Woo specific id.

### How to test the changes in this Pull Request:

1. Disable Jetpack.
1. Add this line to code that runs after `init`. 
```php
print_r( WC_Tracks_Client::get_identity( wp_get_current_user()->ID ) );
```
Here is a good spot
  https://github.com/woocommerce/woocommerce/blob/078057e9b2bbb362be3eddab4009aa116e4e3a1a/includes/tracks/class-wc-site-tracking.php#L58

3. Delete the `tk_ai` cookie from your browser.
3. Refresh the page and see a property `anon [_ui]` printed to the page with an anonymous woo flavored id along with a matching `tk_ai` cookie.
4. Delete the `tk_ai` cookie once again an refresh. Notice the cookie re-appear with the same id.
5. Acitvate Jetpack and see a new Jetpack id associated with the cookie and printed output. *
6. Deactivate Jetpack and see a Jetpack id remain in the cookie and output. **

`*` Jetpack writes over the existing `tk_ai` cookie on activation, even when that cookie was established by a previous connection to Jetpack. This seems like a bug to track separately. 

`**` Jetpack also re-writes the existing `tk_ai` cookie on deactivation. This is possibly a bug related to the one described directly above.

psealock committed:
Tracks: Add a PHP client

psealock commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -98,7 +136,7 @@ public static function get_anon_id() {
 
 			// Did the browser send us a cookie?
-			if ( isset( $_COOKIE['tk_ai'] ) && preg_match( '#^[A-Za-z0-9+/=]{24}$#', $_COOKIE['tk_ai'] ) ) {
+			if ( isset( $_COOKIE['tk_ai'] ) ) {
```
The regex removed here always resulted in `false` for id's generated here and those created by Jetpack:

```
jetpack:oKXKpD6dLQTOxXfdrwjxD2l1
```
this resulted in a new id created every time.

claudiosanches replied:
If this is the format, we probably can run `sanitize_text_field()` to sanitize it.

psealock commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -110,7 +148,7 @@ public static function get_anon_id() {
-				$anon_id = 'jetpack:' . base64_encode( $binary );
+				$anon_id = 'woo:' . base64_encode( $binary );
 
 				if ( ! headers_sent()
```
>Could use `wc_setcookie` helper here. If those other 2 checks are worth (?) you could improve `wc_setcookie` in the process.

~ https://github.com/woocommerce/woocommerce/pull/22625#discussion_r257179911

psealock replied:
>This probably needs some sort of sanitization before returning it because it's saved straight to meta.

in reference to `$anonid` in https://github.com/woocommerce/woocommerce/pull/22625/files#r257180446

psealock commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -86,9 +86,47 @@ public static function build_timestamp() {
+		// If an id is still not found, create one and save it.
+		if ( ! $anon_id ) {
+			$anon_id = self::get_anon_id();
+			update_user_meta( $user_id, 'woo_tracks_anon_id', $anon_id );
```
>https://github.com/woocommerce/woocommerce/pull/22625/files#r257180446

~ https://github.com/woocommerce/woocommerce/pull/22625/files#r257181869

psealock replied:
switched to `update_user_meta` to avoid duplication.

psealock commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -86,9 +86,47 @@ public static function build_timestamp() {
+			update_user_meta( $user_id, 'woo_tracks_anon_id', $anon_id );
+		}
+
+		if ( ! isset( $_COOKIE['tk_ai'] ) && ! headers_sent() ) {
```
>wc_setcookie

~ https://github.com/woocommerce/woocommerce/pull/22625/files#r257182012

psealock commented:
I updated this code to reflect changes requested in the previous PR.

* Use `wc_setcookie`
* Sanitize the `$anonId` before updating meta data using `sanitize_text_field`

There are two checks for `REST_REQUEST` and `XMLRPC_REQUEST` before setting cookies to avoid setting cookies on API requests. I have added an `@todo` while I try to understand the need and their possible inclusion in `wc_setcookies`.

claudiosanches commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -92,10 +92,49 @@ public static function build_timestamp() {
+	public static function get_identity( $user_id ) {
+		if ( class_exists( 'Jetpack' ) ) {
+
+			include_once ABSPATH . 'wp-content/plugins/jetpack/_inc/lib/tracks/client.php';
```
This path may change pending in the WP installation, we should use `WP_PLUGIN_DIR` instead.

```php
include_once trailingslashit( WP_PLUGIN_DIR ) . 'jetpack/_inc/lib/tracks/client.php';
```

claudiosanches commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -92,10 +92,49 @@ public static function build_timestamp() {
+		}
+
+		// Start with a previously set cookie.
+		$anon_id = isset( $_COOKIE['tk_ai'] ) ? wp_unslash( $_COOKIE['tk_ai'] ) : false; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
We probably should not avoid sanitization here.

claudiosanches commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -92,10 +92,49 @@ public static function build_timestamp() {
+		// If an id is still not found, create one and save it.
+		if ( ! $anon_id ) {
+			$anon_id = self::get_anon_id();
+			update_user_meta( $user_id, 'woo_tracks_anon_id', $anon_id );
```
We could rename it to `_woo_tracks_anon_id`, so will be hidden by default, also we probably need to exclude it from our CRUD too: https://github.com/woocommerce/woocommerce/blob/e91f3b88f58a7bcaa4e2a9f771621c7960d9a746/includes/data-stores/class-wc-customer-data-store.php#L25

psealock commented:
Thanks for the review @claudiosanches !

I made changes requested:

* Changed meta key to `_woo_tracks_anon_id` and added it to customer data store for exclusion.
* Updated the `include_once` to Jetpack's get_identity.
* Sanitized and unslashed `$_COOKIE['tk_ai']`

These files now pass the linter.

claudiosanches commented:
@psealock all looks good to me, but I think we don't need a `_woo` prefix, we should use `_woocommerce_tracks_anon_id`, so this will get cleaned by our uninstall script:

https://github.com/woocommerce/woocommerce/blob/master/uninstall.php#L57-L58

psealock committed:
_woocommerce_tracks_anon_id to customere data store

psealock commented:
Thanks @claudiosanches , I switched to `_woocommerce_tracks_anon_id`

claudiosanches commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -92,10 +92,49 @@ public static function build_timestamp() {
+		}
+
+		// Start with a previously set cookie.
+		$anon_id = isset( $_COOKIE['tk_ai'] ) ? wp_unslash( $_COOKIE['tk_ai'] ) : false; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
Just missing `sanitize_text_field` and we are good to go.

psealock replied:
whoops, good call. Fixed now

psealock committed:
add sanitize_text_field

psealock committed:
remove phpcs ignore

mikejolley commented on includes/tracks/class-wc-tracks-client.php:
```diff
@@ -92,10 +92,49 @@ public static function build_timestamp() {
+	 * @return array Identity properties.
+	 */
+	public static function get_identity( $user_id ) {
+		if ( class_exists( 'Jetpack' ) ) {
```
I think I pointed out in another PR, this might not be enough to know if Jetpack is really there. Can we add a file exists on jetpack/_inc/lib/tracks/client.php to be sure before trying to include it?

mikejolley reviewed (approved):
Just 1 note.

claudiosanches committed:
Check if file exists before loading

claudiosanches committed:
Better check for Jetpack

claudiosanches committed:
Clean up

claudiosanches reviewed (approved):
LGTM!
